File: algorithms/strings/urlify/urlify.go
package main

import "fmt"

func urlify(s string, trueLen int) string {
	arr := []byte(s)
	actualIdx := len(arr)

	for i := trueLen - 1; i >= 0; i-- {
		if arr[i] == ' ' {
			arr[actualIdx-1] = '0'
			arr[actualIdx-2] = '2'
			arr[actualIdx-3] = '%'
			actualIdx -= 3
		} else {
			arr[actualIdx-1] = arr[i]
			actualIdx--
		}
	}

	return string(arr)
}

func main() {
	str := "Mr John Smith    "

	encodedStr := urlify(str, 13)
	fmt.Println("Encoded str : ", encodedStr)
}

File: algorithms/strings/urlify/urlify.py
def urlify(str, true_len):
    actual_idx = len(str)
    arr = list(str)

    for i in reversed(range(true_len)):
        if arr[i] == ' ':
            arr[actual_idx-3:actual_idx] = '%20'
            actual_idx -= 3
        else:
            arr[actual_idx - 1] = arr[i]
            actual_idx -= 1

    return ''.join(arr)


str = "Mr John Smith    "
encoded_str = urlify(str, 13)
print("Encoded string : {}".format(encoded_str))
